### ISSUE TITLE
Some question about the position prediction

### ISSUE THREAD

**Issue by Eldo-rado** [MediaBrain-SJTU/MedKLIP#5]
👋  Thanks for your code. I have some question about the 'position prediction'. 

- It is mentioned in the paper that after Transformer, MLPs will predict the existence and position of the entity respectively. 
I can understand the prediction of the existence of the entity, but for the prediction of the position, i wonder, if the entity does not exist in the image, but it has to predict its position, is it reasonable?  What is the significance of the predicted position when the entity does not exist in the image?

- In addition, for position prediction, the anatomic position in X-ray is fixed, so why not directly treat position prediction as a multi-classification task, but do contrastive learning for embedding?

Thank you in advance. 😺

**Comment by chaoyi-wu** [comment]
Your questions are very good！

- Your concern is correct.  We also think it is unreasonable for the model able to predict the position when the entity does not exist in the image. Thus we actually will not calculate position contrastive loss if the entity is labeled as unexist, i.e., the position contrastive loss will only be calculated on the entities labeled as existing. We did not mention this in the paper, mainly because we find it a bit hard to explain this to readers who have not read our code.
- We adopt contrastive learning here, mainly hoping to retain the information contained in the position names. For example, left_lung and right_lung, pointing to two completely opposite locations, their text embeddings will tend to be different, thus helping the network to capture the critical location information quickly. Additions, like right_lung_unspec and right_apical_lung, it is hard to say whether they are overlapped, and their text embeddings will tend to be similar, avoiding misleading the network significantly.

**Comment by Eldo-rado** [comment]
Thank you for your timely reply.

- I got it !

- What you mean is that if you do it as a classification task, each position is independent of each other, like left_lung and right_lung, right_lung_unspec and right_apical_lung, all will be treated as a totally different category.  As the way of embedding, the latter(right_lung_unspec and right_apical_lung) will be closer in the semantic space.  (If I understand correctly, I think it is quite reasonable!)

Thanks again!

**Comment by chaoyi-wu** [comment]
Yes, exactly.

**Comment by Eldo-rado** [comment]
Sorry to bother you again.
I read the paper 'K-Diag' tonight and found that you happen to be the author😌. There are two questions I want to confirm.

- In the paper, $\\boldsymbol{n_i} \in R^{N \times d}, \ \boldsymbol{d_i} \in R^{N \times d}$ . I want to know what do $\boldsymbol{n_i}$ and $\boldsymbol{d_i}$  stand for respectively .
In my understanding, that maybe: $n_i \in R^{1 \times d}, \ d_i \in R^{1 \times d}$  and $(n_i)_{i=1}^N  \in R^{N \times d}, \ \ (d_i) _ {i=1} ^N \in R^{N \times d}$
$n_i$ and $d_i$ are expected to be used instead of  $\\boldsymbol{n_i}$ and $\\boldsymbol{d_i}$. 

- Is the Learnable Prompt Module proposed to transform T into more flexible embedding? Are the set of learnable vectors initialized randomly?

**Comment by chaoyi-wu** [comment]
- Yes. Your understanding is correct. Our notations are indeed a bit confusing here and thank you for pointing it out.
- Yes, we use LPM as an adaptation, here, and they are initialized randomly. Intuitively, we design the LPM motivated by that different pathologies may share some common attributes. You can view the prompt basis as an implicit embedding of different attributes learned by the network. For any pathology, the network all needs to pick a linear combination of the basis (attributes) to form an embedding for it. Thus, we think the module can better enhance the network to capture the shared patterns across pathologies, avoiding overfitting the limited pathology text embedding input.

**Comment by Eldo-rado** [comment]
Hello, after reading the code in detail, I have the following points to confirm.

1. What does `line 158` in `PreTrain_MedKLIP/model_MedKLIP.py` work for?

(Accorfing to the processing in `dataset.py`, I know that smaple_index refers to the position matrix which shape is [75,8].)
```
anatomy_query = self.ana_book[smaple_index,:] # batch, 4 , dim
```
\
2. What does `line 166 & line 167` in `PreTrain_MedKLIP/model_MedKLIP.py` work for?
```
anatomy_query = anatomy_query.reshape(B*Q,8,768)
ll = torch.bmm(anatomy_query, ll ).squeeze()  # B Q 4
```
In my understanding, $ll$ is the position matrix obtained by prediction, and its shape is [B,Q,768].
Why would you do the torch.bmm on $ll$ and anatomy_query? The description in the paper does not seem to cover this operation.
\
3. What are the shapes of $ll$ and cl_labels on line 200?
```
loss_cl = F.cross_entropy(ll,cl_labels)
```

**Comment by chaoyi-wu** [comment]
1. The shape of `smaple_index` is `[B, entity_num, sampled_position_num]`, which is `[B,75,8]` in default setting. Thus the shape of anatomy_query is ```[B, entity_num, sampled_position_num, dim]```, which is `[B,75,8,768]` and `line 158` represents obtaining the embeddings of the sampled positions for each entity query. Note that, during sampling, we always put the positive position embedding at the beginning of the sampled_positions(see `line 81-91` in `dataset/dataset.py`), so the cl_label is always initialized as zero.
2.  `anatomy_query` is the embeddings of the sampled positions and `line 166 & line 167` are calculating the dot product similarity between the position prediction and the sampled position embeddings, just corresponding to the contrastive loss in the paper.
3. the shape of _ll_ is `[B*entity_num, sampled_position_num]`, and cl_labels is `[B*entity_num]`. However, as I mentioned before, the `loss_cl` will only performs on the entities labeled as exist, thus the first dimension `B*entity_num` will be masked with `cl_mask = (labels == 1).squeeze()` (see `line 192&198&199` in `PreTrain_MedKLIP/model_MedKLIP.py`) , so the actual shape of _ll_ and cl_labels will depend on how much entities are labeled as exist.

**Comment by Eldo-rado** [comment]
Thank you. I understand now.
I didn't understand it because I was wondering what the '4' in the comment meant, how can I get this '4'.😂
<img width="772" alt="image" src="https://user-images.githubusercontent.com/128444206/231632652-bc6b7d1a-5b96-4d47-af7f-5e2cdbb46954.png">

**Comment by chaoyi-wu** [comment]
4 corresponds to the `sampled_postion_num` and we sampled 4 positions for cl_loss at the beginning. The comments were written at that moment while it equals 8 in this version. I have modified these comments. 

If you want to change the `sampled_postion_num` to other numbers, you can refer to `line 87 in dataset/dataset.py`, `temp_list = temp_list + random.sample(np.where(class_label != 1)[0].tolist(),7)`, modifying the number 7 to any other numbers like 3 or 15, that will sample 4 or 16 sampled positions.
